core/crypto/crypto_store/crypto_store_impl.hpp

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_CRYPTO_STORE_IMPL_HPP
+#define KAGOME_CRYPTO_STORE_IMPL_HPP
+
+#include <boost/variant.hpp>
+#include "common/blob.hpp"
+#include "crypto/bip39/bip39_provider.hpp"
+#include "crypto/crypto_store.hpp"
+#include "crypto/ed25519_provider.hpp"
+#include "crypto/secp256k1_provider.hpp"
+#include "crypto/sr25519_provider.hpp"
+
+namespace kagome::crypto {
+
+  // TODO (yuraz): PRE-446 ECDSA will be added later
+  namespace store {
+    using KeyPair = boost::variant<ED25519Keypair, SR25519Keypair>;
+    using PublicKey = common::Blob<32>;
+    using Seed = common::Hash256;
+
+    enum class CryptoId { ED25519, SR25519, SECP256k1 };
+  }  // namespace store
+
+  enum class CryptoStoreError {
+    WRONG_KEYFILE_NAME = 1,
+    UNSUPPORTED_KEY_TYPE,
+    UNSUPPORTED_CRYPTO_TYPE,
+    NOT_REGULAR_FILE,
+    FAILED_OPEN_FILE,
+    INVALID_FILE_FORMAT,
+    INCONSISTENT_KEYFILE,
+    KEYS_PATH_IS_NOT_DIRECTORY,
+    WRONG_SEED_SIZE
+  };
+
+  class CryptoStoreImpl : public CryptoStore {
+   public:
+    ~CryptoStoreImpl() override = default;
+
+    CryptoStoreImpl(std::shared_ptr<ED25519Provider> ed25519_provider,
+                    std::shared_ptr<SR25519Provider> sr25519_provider,
+                    std::shared_ptr<Secp256k1Provider> secp256k1_provider,
+                    std::shared_ptr<Bip39Provider> bip39_provider);
+
+    /**
+     * @brief sets keys directory up and checks that it is correct
+     */
+    outcome::result<void> initialize(
+        std::filesystem::path keys_directory) override;
```

Comment:
Why not to pass keys directory to the constructor?

Reply:
keys directory can be invalid, in this case we need to indicate the error and the only way of doing that in constructor is throwing an exception, but it's not a good idea.

Reply:
directory path will be checked outside, so you don't need to check anything